### ISSUE TITLE
Cannot read property 'target' of null

### ISSUE THREAD

**Issue by syuilo** [npm/npm#10686]
`npm install <any package name>` Then an error occurs.

```
npm ERR! Linux 4.2.5-1-ARCH
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install"
npm ERR! node v5.1.1
npm ERR! npm  v3.5.2

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /var/misskey-web/node_modules/misskey-web/npm-debug.log
```

Full log here:
https://gist.github.com/syuilo/837f61e36657d1e11458

**error occurs only in the directory that include "package.json".**

example of my package.json:

```
{
  "name": "misskey-web",
  "version": "0.0.0",
  "description": "Misskey Web",
  "main": "./built/index.js",
  "repository": {
    "url": "https://github.com/MissKernel/Misskey-Web.git"
  },
  "scripts": {
    "start-production": "set NODE_ENV=production node ./built/index.js",
    "start": "node ./built/index.js",
    "dtsm": "dtsm --insight false install",
    "build": "gulp build",
    "watch": "gulp watch",
    "lint": "gulp lint",
    "clean": "gulp clean",
    "clean-all": "gulp clean-all",
    "test": "gulp test"
  },
  "license": "MIT",
  "browserify": {
    "transform": [
      "debowerify",
      [
        "jadeify",
        {
          "compileDebug": false,
          "pretty": false
        }
      ]
    ]
  },
  "devDependencies": {
    "babel-preset-es2015": "6.1.18",
    "bower": "1.6.8",
    "browserify": "12.0.1",
    "debowerify": "1.3.1",
    "dtsm": "0.13.0",
    "event-stream": "3.3.2",
    "glob": "6.0.1",
    "gulp": "3.9.0",
    "gulp-babel": "6.1.1",
    "gulp-less": "3.0.5",
    "gulp-livescript": "2.4.0",
    "gulp-minify-css": "1.2.1",
    "gulp-notify": "2.2.0",
    "gulp-plumber": "1.0.1",
    "gulp-tslint": "3.6.0",
    "gulp-typescript": "2.9.2",
    "gulp-uglify": "1.5.1",
    "jadeify": "^4.5.0",
    "livescript": "1.4.0",
    "tslint": "3.1.1",
    "typescript": "1.7.3",
    "typescript-require": "0.2.9-1",
    "vinyl-buffer": "^1.0.0",
    "vinyl-source-stream": "1.1.0"
  },
  "dependencies": {
    "body-parser": "1.14.1",
    "cheerio-httpcli": "0.3.8",
    "compression": "1.6.0",
    "connect-mongo": "0.8.2",
    "cookie": "0.2.3",
    "cookie-parser": "1.4.0",
    "cors": "2.7.1",
    "escape-html": "1.0.3",
    "express": "4.13.3",
    "express-minify": "0.1.6",
    "express-session": "1.12.1",
    "gm": "1.21.1",
    "jade": "1.11.0",
    "jquery": "2.1.4",
    "jquery.transit": "0.9.12",
    "less": "2.5.3",
    "log-cool": "1.0.1",
    "marked": "0.3.5",
    "moment": "2.10.6",
    "mongoose": "4.2.9",
    "multer": "1.1.0",
    "redis": "2.4.2",
    "request": "2.67.0",
    "socket.io": "1.3.7",
    "subdomain": "1.2.0",
    "three": "0.73.0",
    "vhost": "3.0.2"
  },
  "private": true
}
```

---

OS ArchLinux
Node 5.1.1
npm 3.5.2

---

Sorry my English. Thanks.

**Comment by othiym23** [comment]
This is a great bug report – thanks for both the log and the `package.json`. We'll take a look at it very soon!

**Comment by syuilo** [comment]
Thanks for reply!

Where the error occurred is like here:
https://github.com/npm/npm/blob/master/lib/install/filter-invalid-actions.js#L24

Good luck!

**Comment by othiym23** [comment]
I wasn't able to reproduce this from a blank slate (which makes it great that you included `npm-debug.log` – thanks!). This will make things a little trickier to track down, but it indicates that you can probably install successfully by running `rm -rf node_modules && npm install` in your application. Let us know if that doesn't work!

**Comment by syuilo** [comment]
I tried it, but the same error occurs.

**Comment by othiym23** [comment]
Thanks for following up! From the debug log, it looks like you were installing `misskey-web` for the first time when npm crashed. Is that right? If so, the fact that you and I are getting different outcomes means there's probably a data problem of some kind. It would be helpful if you tried running `npm cache clean`, but I can understand if you didn't want to have to rebuild your npm cache.

**Comment by syuilo** [comment]
The ``npm cache clean`` was also the result after did not change.

Note:
Error in local(Windows) does not occur.
Error will occur only Arch(VPS).

**Comment by othiym23** [comment]
Thanks for that information! I tried to replicate the failure on an Ubuntu trusty host running Node.js 5.1.1 and `npm@3.5.2`, and I was still able to install all of the dependencies successfully. This still feels like a data issue to me, but I'm uncertain where the bad data could be coming from. I've talked to @iarna about how `pkg.parent` could be `null`, and the only way it could happen would be if somehow an action were being taken on the topmost dependency, which should never happen (and doesn't seem to be indicated in the logs, either).

Does the failure still happen with `npm@3.5.1`, `npm@3.5.0`, and `npm@3.3.12`? (Please don't try with `npm@3.4.0` or `npm@3.4.1`, because they had a problem that makes it difficult to use them to reinstall npm after you've installed them.)

**Comment by syuilo** [comment]
As a result of try to 3.3.12, I were able to successfully installed!
Awesome, thank you very much!!

**Comment by kenokabe** [comment]
Hi, 
I have the same issue:

```
$ npm install timeengine
```

timeengine(my own npm library) package.json

```
  "devDependencies": {
    "babel-cli": "^6.3.15",
    "babel-preset-es2015": "^6.3.13",
    "babel-preset-react": "^6.3.13",
    "eslint": "^1.10.3"
  },
```

When to install the devDependencies, the following error occurs:

```
$ npm install
npm ERR! Linux 4.3.0-1-MANJARO
npm ERR! argv "/home/ken/.nvm/versions/node/v5.1.1/bin/node" "/usr/bin/npm" "install"
npm ERR! node v5.1.1
npm ERR! npm  v3.5.1

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ken/projects/node_modules/timeengine/npm-debug.log

```

**Comment by smartmouse** [comment]
This may be a regression?  npm install timeengine  works fine on Node v5.1.0  and NPM 3.3.12.

**Comment by kenokabe** [comment]
`npm install timeengine` works. yes.

To install  "devDependencies", in `/path/timeengine`, 
`npm install` occurs the error. Thanks.

**Comment by smartmouse** [comment]
This also works for me,

https://gist.github.com/smartmouse/58129b2ec6b0ca040ef7

**Comment by kenokabe** [comment]
I double-checked, and the got the same error.
Probably because my node/npm version is newer.

```
npm ERR! node v5.1.1
npm ERR! npm  v3.5.1
```

**Comment by damadorPL** [comment]
36105 verbose stack TypeError: Cannot read property 'target' of null
36105 verbose stack     at module.exports (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\lib\install\filter-invalid-actions.js:24:35)
36105 verbose stack     at Array.<anonymous> (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
36105 verbose stack     at LOOP (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
36105 verbose stack     at C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:18:7
36105 verbose stack     at Object.Installer.computeLinked (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\lib\install.js:411:41)
36105 verbose stack     at Array.<anonymous> (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
36105 verbose stack     at LOOP (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
36105 verbose stack     at C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:18:7
36105 verbose stack     at module.exports (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\lib\install\diff-trees.js:50:3)
36105 verbose stack     at Array.<anonymous> (C:\Users\damador\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
36106 verbose cwd C:\Users\damador\AppData\Roaming\npm\node_modules\bootstrap
36107 error Windows_NT 10.0.10586
36108 error argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\damador\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install"
36109 error node v5.3.0
36110 error npm  v3.5.3
36111 error Cannot read property 'target' of null
36112 error If you need help, you may report this error at:
36112 error     <https://github.com/npm/npm/issues>
36113 verbose exit [ 1, true ]

{
  "_args": [
    [
      "bootstrap",
      ""
    ]
  ],
  "_from": "bootstrap@latest",
  "_id": "bootstrap@3.3.6",
  "_inCache": true,
  "_installable": true,
  "_location": "/bootstrap",
  "_nodeVersion": "0.12.7",
  "_npmUser": {
    "email": "getbootstrap@gmail.com",
    "name": "twbs"
  },
  "_npmVersion": "2.11.3",
  "_phantomChildren": {},
  "_requested": {
    "name": "bootstrap",
    "raw": "bootstrap",
    "rawSpec": "",
    "scope": null,
    "spec": "latest",
    "type": "tag"
  },
  "_requiredBy": [
    "#USER"
  ],
  "_resolved": "https://registry.npmjs.org/bootstrap/-/bootstrap-3.3.6.tgz",
  "_shasum": "8de8f749dc8a743f2ac5b510d988371e3daa6589",
  "_shrinkwrap": null,
  "_spec": "bootstrap",
  "_where": "",
  "author": {
    "name": "Twitter, Inc."
  },
  "bugs": {
    "url": "https://github.com/twbs/bootstrap/issues"
  },
  "dependencies": {},
  "description": "The most popular front-end framework for developing responsive, mobile first projects on the web.",
  "devDependencies": {
    "btoa": "~1.1.2",
    "glob": "~6.0.1",
    "grunt": "~0.4.5",
    "grunt-autoprefixer": "~3.0.3",
    "grunt-contrib-clean": "~0.6.0",
    "grunt-contrib-compress": "~0.14.0",
    "grunt-contrib-concat": "~0.5.1",
    "grunt-contrib-connect": "~0.11.2",
    "grunt-contrib-copy": "~0.8.0",
    "grunt-contrib-csslint": "~0.5.0",
    "grunt-contrib-cssmin": "~0.14.0",
    "grunt-contrib-htmlmin": "~0.6.0",
    "grunt-contrib-jade": "~0.15.0",
    "grunt-contrib-jshint": "~0.11.2",
    "grunt-contrib-less": "~1.1.0",
    "grunt-contrib-qunit": "~0.7.0",
    "grunt-contrib-uglify": "~0.9.2",
    "grunt-contrib-watch": "~0.6.1",
    "grunt-csscomb": "~3.1.0",
    "grunt-exec": "~0.4.6",
    "grunt-html": "~5.0.1",
    "grunt-jekyll": "~0.4.2",
    "grunt-jscs": "~2.3.0",
    "grunt-saucelabs": "~8.6.1",
    "grunt-sed": "github:twbs/grunt-sed#v0.2.0",
    "load-grunt-tasks": "~3.3.0",
    "markdown-it": "^5.0.0",
    "npm-shrinkwrap": "^200.4.0",
    "time-grunt": "^1.2.1"
  },
  "directories": {},
  "dist": {
    "shasum": "8de8f749dc8a743f2ac5b510d988371e3daa6589",
    "tarball": "http://registry.npmjs.org/bootstrap/-/bootstrap-3.3.6.tgz"
  },
  "engines": {
    "node": ">=0.10.1"
  },
  "files": [
    "Gruntfile.js",
    "LICENSE",
    "dist",
    "fonts",
    "grunt",
    "js/*.js",
    "less/**/*.less"
  ],
  "gitHead": "81df608a40bf0629a1dc08e584849bb1e43e0b7a",
  "homepage": "http://getbootstrap.com",
  "jspm": {
    "files": [
      "css",
      "fonts",
      "js"
    ],
    "main": "js/bootstrap",
    "shim": {
      "js/bootstrap": {
        "deps": "jquery",
        "exports": "$"
      }
    }
  },
  "keywords": [
    "css",
    "framework",
    "front-end",
    "less",
    "mobile-first",
    "responsive",
    "web"
  ],
  "less": "less/bootstrap.less",
  "license": "MIT",
  "main": "./dist/js/npm",
  "maintainers": [
    {
      "name": "twbs",
      "email": "bigj95t+bsnpm@gmail.com"
    }
  ],
  "name": "bootstrap",
  "optionalDependencies": {},
  "readme": "ERROR: No README data found!",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/twbs/bootstrap.git"
  },
  "scripts": {
    "test": "grunt test"
  },
  "style": "dist/css/bootstrap.css",
  "version": "3.3.6"
}

**Comment by Tanato** [comment]
:+1:

**Comment by georaldc** [comment]
Exact same error, though unfortunately I do not have log files to provide at the moment. Basically trying to follow roaks3's suggestion here to force a specific version of yeoman-environment:

https://github.com/yeoman/generator-angular/issues/1239#issuecomment-167497418

But running npm install just gives me the property 'target' of null error

**Comment by skleeschulte** [comment]
Same error here when trying to install dev dependencies of jquery 2.1.4.

Environment:
 * Windows 7 x64
 * node v4.2.4
 * npm v3.5.2

Steps to reproduce (I used `cmd` in empty directory `C:\dev\jquery`):

    npm install jquery
    xcopy "node_modules/jquery" /s /e
    rmdir node_modules /s /q
    npm install

Deleting `%AppData%\npm-cache` did not help.

**npm-debug.log**: [npm-debug.log.txt](https://github.com/npm/npm/files/77850/npm-debug.log.txt)

**Comment by iarna** [comment]
This is super disturbing because it means an action was generated for a node w/o a parent, which ordinarily indicates the top of the install tree, eg, your project, and it makes no sense for npm to be trying to install or remove it.

I think the next step in this will be to add a validation step that scans the actions for this scenario and errors out with (hopefully) more useful information about what's happening / how this is happening.

**Comment by tonybranfort** [comment]
Same error.  Works fine on versions of npm < 3.5. Error begins with 3.5.0. 

```shell
[~]$ npm install apom
[~/node_modules/apom]$ npm install 

npm ERR! Linux 4.1.10-17.31.amzn1.x86_64
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v4.2.4
npm ERR! npm  v3.5.0

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ec2-user/node_modules/apom/npm-debug.log


```
I tried on these versions: 
* Fails with node v4.2.4 / npm v3.5.4
* Fails with node v4.2.4 / npm v3.5.2
* Fails with node v4.2.4 / npm v3.5.0
* Fails with node v5.4.1 / npm v3.5.4
* Successful with node 4.2.4 / npm 3.1.3
* Successful with node 4.2.4 / npm 3.2.2
* Successful with node 4.2.4 / npm 3.4.1
* Successful with node v0.12.7 / npm v2.11.3


All were installed on a new AWS EC2 Linux t2.small build image ami-f0091d91.

**Comment by jadbox** [comment]
npm  v3.7.1

Same issue: 
```npm install juttle-elastic-adapter@^0.4.0                                                                         [11:34:09]
npm ERR! Linux 4.2.0-0.bpo.1-amd64
npm ERR! argv "/usr/local/bin/node" "/home/jdunlap/npm/bin/npm" "install" "juttle-elastic-adapter@^0.4.0"
npm ERR! node v5.5.0
npm ERR! npm  v3.7.1

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/jdunlap/npm/lib/node_modules/outrigger/npm-debug.log```

**Comment by Shulin** [comment]
I'm actually getting this too. But it's happening when I'm executing a 'postinstall' script for a project that is in turn using child process to execute an 'npm install' of another package.

Currently happening on 
node v4.2.4
npm  v3.5.0

Works fine on
node v0.10.31
npm v1.4.23

npm error:

> npm ERR! Linux 3.16.0-4-amd64
> npm ERR! argv "/usr/local/nodejs-v4.2.4-54/bin/node" "/usr/local/nodejs-v4.2.4/bin/npm" "install" "my-test-dependent-project@0.2.6"

**Comment by jadbox** [comment]
Same issue here with 3.7.1, even after deleting all my cache. Going back to 3.3.12 doesn't have the problem.

**Comment by noderat** [comment]
I'm having this issue occur as well, but I've noticed I'm also running into low memory issues on the aws instance around the same time as the npm install.

**Comment by tonybranfort** [comment]
Interesting...
> @noderat ...I'm also running into low memory issues on the aws instance...

I just installed node v4.2.4 and npm v3.5.4 on a fresh, new aws t2.micro instance to see if that might be an issue.  After those installs plus my own 'npm install apom', 
```
MemTotal:        1019452 kB
MemFree:          602168 kB
MemAvailable:     970592 kB
```
And I still got the error 
```
$npm install
npm ERR! Linux 4.1.10-17.31.amzn1.x86_64
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v4.2.4
npm ERR! npm  v3.5.4

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ec2-user/node_modules/apom/npm-debug.log
```

**Comment by syuilo** [comment]
RAM of My VPS in which the error occurred is 1GB.

**Comment by tonybranfort** [comment]
aws t2.medium, 4 GB, 
```
MemTotal:        4049992 kB
MemFree:         3919912 kB
MemAvailable:    3983120 kB
```
still same error...
```
npm ERR! Linux 4.1.13-19.31.amzn1.x86_64
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v4.2.4
npm ERR! npm  v3.5.4

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ec2-user/node_modules/apom/npm-debug.log

```

**Comment by fahrvergnuugen** [comment]
Getting this error too when trying to update my already installed packages. Is there a workaround?

**Comment by jadbox** [comment]
@fahrvergnuugen I haven't found any... had to degrade back to npm 3.3.12 stable (be sure to set -progress=false on installs).

**Comment by mojodna** [comment]
My repro:

```bash
$ mkdir -p /tmp/null-target
$ cd /tmp/null-target
$ npm install inherits # arbitrarily chosen package w/ no deps
$ cd node_modules/inherits
$ npm install is-property # arbitrarily chosen package w/ no deps
npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/seth/.nvm/versions/node/v5.6.0/bin/node" "/Users/seth/.nvm/versions/node/v5.6.0/bin/npm" "install" "is-property"
npm ERR! node v5.6.0
npm ERR! npm  v3.6.0

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /private/tmp/node_modules/inherits/npm-debug.log
```

Stack trace (from `npm-debug.log`):

```
TypeError: Cannot read property 'target' of null
    at module.exports (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/lib/install/filter-invalid-actions.js:24:35)
    at Array.<anonymous> (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/bind-actor.js:15:8)
    at LOOP (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/chain.js:15:14)
    at /Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/chain.js:18:7
    at Object.Installer.computeLinked (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/lib/install.js:411:41)
    at Array.<anonymous> (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/bind-actor.js:15:8)
    at LOOP (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/chain.js:15:14)
    at /Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/chain.js:18:7
    at module.exports (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/lib/install/diff-trees.js:50:3)
    at Array.<anonymous> (/Users/seth/.nvm/versions/node/v5.6.0/lib/node_modules/npm/node_modules/slide/lib/bind-actor.js:15:8)
```

Candidate patch:

```diff
diff --git i/lib/install/filter-invalid-actions.js w/lib/install/filter-invalid-actions.js
index f90bf0b..6ca63dc 100644
--- i/lib/install/filter-invalid-actions.js
+++ w/lib/install/filter-invalid-actions.js
@@ -21,9 +21,9 @@ module.exports = function (top, differences, next) {
   while (action = differences.shift()) {
     var cmd = action[0]
     var pkg = action[1]
-    if (pkg.isInLink || pkg.parent.target || pkg.parent.isLink) {
+    if (pkg.isInLink || pkg.parent == null || pkg.parent.target || pkg.parent.isLink) {
       // we want to skip warning if this is a child of another module that we're removing
-      if (!pkg.parent.removing) {
+      if (pkg.parent && !pkg.parent.removing) {
         log.warn('skippingAction', 'Module is inside a symlinked module: not running ' +
           cmd + ' ' + packageId(pkg) + ' ' + path.relative(top, pkg.path))
       }
```

**Comment by kaizhu256** [comment]
my travis-ci build just failed npm-install today for no reason (and its only one package dependency) @ https://travis-ci.org/kaizhu256/node-utility2/jobs/108725798

```
[MODE_BUILD=npmTestPublished] - 2016-02-12T07:11:03Z - npm-testing published package utility2
/tmp/app
└── utility2@2016.1.1 
npm WARN enoent ENOENT: no such file or directory, open '/tmp/app/package.json'
npm WARN app No description
npm WARN app No repository field.
npm WARN app No README data
npm WARN app No license field.
npm ERR! Linux 3.13.0-40-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v5.6.0/bin/node" "/home/travis/.nvm/versions/node/v5.6.0/bin/npm" "install"
npm ERR! node v5.6.0
npm ERR! npm  v3.6.0
npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/app/node_modules/utility2/npm-debug.log
npm ERR! Linux 3.13.0-40-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v5.6.0/bin/node" "/home/travis/.nvm/versions/node/v5.6.0/bin/npm" "run-script" "build-ci"
npm ERR! node v5.6.0
npm ERR! npm  v3.6.0
npm ERR! code ELIFECYCLE
npm ERR! utility2@2016.1.2 build-ci: `./index.sh shRun shReadmeBuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the utility2@2016.1.2 build-ci script './index.sh shRun shReadmeBuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the utility2 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ./index.sh shRun shReadmeBuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs utility2
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls utility2
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/kaizhu256/node-utility2/npm-debug.log
The command "npm run-script build-ci" exited with 1.
Done. Your build exited with 1.
```

**Comment by kaizhu256** [comment]
@iarna , after debugging https://github.com/npm/npm/blob/master/lib/install/filter-invalid-actions.js, i found that its being called in the top of the install tree

i can reliably reproduce the error inside docker image node:latest

1. npm install <foo> # foo can be any package
2. cd node_modules/<foo>
3. npm install # fails because current directory (and module) node_modules/<foo> has no parent


here's the shell output when i npm install lodash, and then run 'npm install' again inside node_modules/lodash

```
(node2)root@42fb0b51dbae:/# mkdir -p /tmp/app
(node2)root@42fb0b51dbae:/# cd /tmp/app



(node2)root@42fb0b51dbae:/tmp/app# npm install lodash # step 1
npm info it worked if it ends with ok
npm info using npm@3.6.0
npm info using node@v5.6.0
npm info attempt registry request try #1 at 9:15:50 AM
npm http request GET https://registry.npmjs.org/lodash
npm http 200 https://registry.npmjs.org/lodash
npm info addNameTag [ 'lodash', 'latest' ]
npm info retry fetch attempt 1 at 9:15:50 AM
npm info attempt registry request try #1 at 9:15:50 AM
npm http fetch GET https://registry.npmjs.org/lodash/-/lodash-4.3.0.tgz
npm http fetch 200 https://registry.npmjs.org/lodash/-/lodash-4.3.0.tgz
npm info lifecycle lodash@4.3.0~preinstall: lodash@4.3.0
npm info linkStuff lodash@4.3.0
npm info lifecycle lodash@4.3.0~install: lodash@4.3.0
npm info lifecycle lodash@4.3.0~postinstall: lodash@4.3.0
/tmp/app
`-- lodash@4.3.0 

npm WARN enoent ENOENT: no such file or directory, open '/tmp/app/package.json'
npm WARN app No description
npm WARN app No repository field.
npm WARN app No README data
npm WARN app No license field.
npm info ok 



(node2)root@42fb0b51dbae:/tmp/app# cd node_modules/lodash/ # step 2



(node2)root@42fb0b51dbae:/tmp/app/node_modules/lodash# npm install # step 3
npm info it worked if it ends with ok
npm info using npm@3.6.0
npm info using node@v5.6.0
npm ERR! Linux 4.1.5-x86_64-linode61
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v5.6.0
npm ERR! npm  v3.6.0

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/app/node_modules/lodash/npm-debug.log
```

**Comment by kaizhu256** [comment]
@othiym23 , the use-case for running npm-install inside a topmost dependency is to run npm-test after i publish a package.

in my case, after i publish the package utility2,
i run:
```npm install utility2 && cd node_modules/utility2 && npm install && npm test```  as a post-publish test in travis-ci

**Comment by kaizhu256** [comment]
this was my [hacky] shell-script workaround to run npm-test on an npm-published package
```
npm install foo
cd node_modules/foo
sed -in -e 's/ "_requiredBy":/ "_requiredBy_":/' package.json # npm-install hack
npm install # npm-install devDependencies in published package
npm test # npm-test published package
```

**Comment by MadWombat** [comment]
Similar issue here
<pre>
 wombat  ~  npm outdated -g

> npm@3.7.4 prepublish /usr/local/lib/node_modules/npm
> node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 doc

scripts/doc-build.sh doc/cli/npm-access.md man/man1/npm-access.1
scripts/doc-build.sh doc/cli/npm-adduser.md man/man1/npm-adduser.1
scripts/doc-build.sh doc/cli/npm-bin.md man/man1/npm-bin.1
scripts/doc-build.sh doc/cli/npm-bugs.md man/man1/npm-bugs.1
node cli.js install marked-man --no-global
npm ERR! Darwin 15.3.0
npm ERR! argv "/usr/local/Cellar/node/5.6.0/bin/node" "/usr/local/lib/node_modules/npm/cli.js" "install" "marked-man" "--no-global"
npm ERR! node v5.6.0
npm ERR! npm  v3.7.4

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/local/lib/node_modules/npm/npm-debug.log
make[1]: *** [node_modules/.bin/marked-man] Error 1
make: *** [man/man1/npm-adduser.1] Error 2
make: *** Waiting for unfinished jobs....
</pre>

**Comment by prettydiff** [comment]
I can produce this with prettydiff.  Steps to produce:

    npm install prettydiff
    cd node_modules/prettydiff
    npm install

I get this error:

    npm ERR! Darwin 14.5.0
    npm ERR! argv "/Users/echeney/.nvm/versions/node/v5.6.0/bin/node" "/Users/echeney/.nvm/versions/node/v5.6.0/bin/npm" "install"
    npm ERR! node v5.6.0
    npm ERR! npm  v3.6.0
    
    npm ERR! Cannot read property 'target' of null
    npm ERR! 
    npm ERR! If you need help, you may report this error at:
    npm ERR!     <https://github.com/npm/npm/issues>
    
    npm ERR! Please include the following file with any support request:
    npm ERR!     /Users/echeney/node_modules/prettydiff/npm-debug.log

The error condition is the same whether I install locally, globally, or locally into a different location.

**Comment by blazejgrzelinski** [comment]
I have got this problem too:
npm ERR! Darwin 15.0.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v5.5.0
npm ERR! npm  v3.7.3

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/blazejgrzelinski/projects/hyphen-test2/node_modules/hyphen-js/npm-debug.log

and for me helped removing this three lines from package.json

**  "_requiredBy": [
    "#USER"
  ],**

For sure problem is connected to npm and node version but I would avoid reinstalling it and playing with different version.

**Comment by LynnOwens** [comment]
>  and for me helped removing this three lines from package.json
> 
> ** "_requiredBy": [
> "#USER"
> ],**

Same.  Windows 10 btw.  Current node.js and npm.  got this error installing bootstrap

**Comment by frozensnowy** [comment]
When doing these steps:
`npm install git://github.com/seishun/node-steam-resources.git
cd node_modules/steam-resources
npm install steam`

I get the exact same error:
`npm ERR! Windows_NT 10.0.10240
npm ERR! argv "C:\\Node\\node.exe" "C:\\Node\\node_modules\\npm\\bin\\npm-cli.js" "install" "steam"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0

npm ERR! Cannot read property 'target' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     C:\cleverbot-steam-master\node_modules\steam-resources\npm-debug.log`

I'm kinda new to this, but, seems like I'm not the only one with the issue?

**Comment by frozensnowy** [comment]
> and for me helped removing this three lines from package.json
> 
> ** "_requiredBy": [
> "#USER"
> ],**

These don't seem to be in my package.json, yet it fails. Any way around this? 
(I could be looking wrong, since I'm kind of new to this)

**Comment by prettydiff** [comment]
The interim solution begs a question.  If the problematic data is dynamically generated how do we prevent it from reappearing after it is manually deleted?  My concern is for my users who may wish to download my package and run the unit tests (via NPM scripts), but are blocked by this error.

**Comment by ryanmolden** [comment]
Also was able to bypass by removing the requiredBy lines like @LynnOwens

**Comment by 8026787** [comment]
Same problem.
Full log is here: https://gist.github.com/8026787/4bfcb11246151240fcd5

**Comment by othiym23** [comment]
This may be "fixed" by #11947 (there may be a deeper bug, but npm should no longer crash with this error), which is included in `npm@3.8.3`, so if those of you have encountered this bug could give `npm@next` a look and let us know if the behavior changes, that would be very useful! Thanks!

**Comment by fiws** [comment]
I have the same problem trying to install `node-dht-sensor` on my Raspberry Pi 2. npm@3.8.3 does not solve the problem for me @othiym23 .

## Terminal output
```
λ tau /tmp → npm i node-dht-sensor

> node-dht-sensor@0.0.11 preinstall /tmp/node_modules/.staging/node-dht-sensor-7ad49927
> ./preinst

Library bcm2835 found.
Installing Native Abstractions for Node.js (NAN)...
npm ERR! Linux 4.1.19-4-ARCH
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "nan"
npm ERR! node v5.9.0
npm ERR! npm  v3.8.2

npm ERR! Cannot read property 'target' of null
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/node_modules/.staging/node-dht-sensor-7ad49927/npm-debug.log
Rebuiling addon...
module.js:341
    throw err;
    ^

Error: Cannot find module 'nan'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at [eval]:1:1
    at Object.exports.runInThisContext (vm.js:54:17)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:413:34)
    at node.js:550:27
    at _combinedTickCallback (node.js:376:9)
gyp: Call to 'node -e "require('nan')"' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/home/fiws/npm/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:305:16)
gyp ERR! stack     at emitTwo (events.js:100:13)
gyp ERR! stack     at ChildProcess.emit (events.js:185:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:204:12)
gyp ERR! System Linux 4.1.19-4-ARCH
gyp ERR! command "/usr/bin/node" "/home/fiws/npm/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/node_modules/.staging/node-dht-sensor-7ad49927
gyp ERR! node -v v5.9.0
gyp ERR! node-gyp -v v3.3.1
gyp ERR! not ok 
npm WARN enoent ENOENT: no such file or directory, open '/tmp/package.json'
npm WARN tmp No description
npm WARN tmp No repository field.
npm WARN tmp No README data
npm WARN tmp No license field.
npm ERR! Linux 4.1.19-4-ARCH
npm ERR! argv "/usr/bin/node" "/home/fiws/npm/bin/npm" "i" "node-dht-sensor"
npm ERR! node v5.9.0
npm ERR! npm  v3.8.3
npm ERR! code ELIFECYCLE

npm ERR! node-dht-sensor@0.0.11 preinstall: `./preinst`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-dht-sensor@0.0.11 preinstall script './preinst'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the node-dht-sensor package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ./preinst
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs node-dht-sensor
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls node-dht-sensor
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /tmp/npm-debug.log
λ tau /tmp → npm -v
3.8.3
```

I noticed that `npm install nan` did not use the npm in my path. I'm not sure if this is a bug.

### `npm-debug.log`

https://gist.github.com/fiws/31911cef39bdc2bc7d66

**Comment by othiym23** [comment]
@fiws the fact that the failing install is using a different npm is pertinent I. this case, because it's using `npm@3.8.2`, which doesn't have the patch I mentioned. Can you get things set up so `npm@3.8.3` is used for both the parent and child installs?

**Comment by fiws** [comment]
@othiym23 I renamed by `/usr/bin/npm` temporarily.
Now I'm in an install loop
```
λ tau /tmp → npm i node-dht-sensor               

> node-dht-sensor@0.0.11 preinstall /tmp/node_modules/.staging/node-dht-sensor-7ad49927
> ./preinst

Library bcm2835 found.
Installing Native Abstractions for Node.js (NAN)...

> node-dht-sensor@0.0.11 preinstall /tmp/node_modules/.staging/node-dht-sensor-7ad49927/node_modules/.staging/node-dht-sensor-da36ecd2
> ./preinst

Library bcm2835 found.
Installing Native Abstractions for Node.js (NAN)...

> node-dht-sensor@0.0.11 preinstall /tmp/node_modules/.staging/node-dht-sensor-7ad49927/node_modules/.staging/node-dht-sensor-da36ecd2/node_modules/.staging/node-dht-sensor-a4e2a7a3
> ./preinst

[continues forever]
```

But hey, no error :sweat_smile:

**Comment by othiym23** [comment]
Thanks, @fiws, that's useful information to have. Does this failure occur in more than one environment? Sit would be even more useful to get a repro case nailed down.

**Comment by fiws** [comment]
@othiym23 i created a [test package](https://github.com/fiws/test-npm-bug-10686) that reproduces the `Cannot read property 'target' of null` on v3.8.2 and also has the install loop on v3.8.3
```
npm i fiws/test-npm-bug-10686 
```

I tested this on my raspberry pi and my Arch Linux Desktop, same results.

**Comment by ackalker** [comment]
@othiym23 & others

I have bisected the problem on Arch Linux, which currently has node version 5.9.0, npm version 3.8.2 (which has the problem) installed.
First known bad commit appears to be:
f5f166fa352d493e99be9c1c7dd0a21fbbd53748 `install: directlyRequested -> userRequired`.
Last known good commit (direct parent of above) appears to be:
7efd3bc5ec77f54b606c75cfd5f01122d49676e9 `install: Emit install warnings using error-handler`

I have manually verified the last known good and first known bad commits, the bad one produces the error, the good one doesn't.
As test case I used issue #11941, using the following steps in a freshly checked out repository:

```sh
$ git bisect start HEAD v3.0.0
$ git bisect run ${HOME}/bisect-npm
```
The script `[bisect-npm](https://gist.github.com/5e97d0ab3f9311cf70fd)` is a slightly adapted version of the script which I use to setup a set of global npm modules on a new system. Please feel free to use it to reproduce, perhaps even use as a simple test harness for issues like this.

**Comment by ackalker** [comment]
[Full log](https://gist.github.com/1f40da305eb4a7b2b214) (warning: big file) of the git bisect session`

**Comment by ackalker** [comment]
I'm sorry to say, using `npm` installed from commit 473d3244a8ddfd6b260d0aa0d395b119d595bf97 gives a [different error](https://gist.github.com/a6471b06f6d36073d20f) (same with HEAD @ f2ae3ba).
Please see my bisection result in https://github.com/npm/npm/issues/10686#issuecomment-200353028 , the problem may be elsewhere.

**Comment by ibc** [comment]
This happens to me in a Linux Debian Stretch virtual host with just RAM 512MB.
It does work in another Debian Stretch (non virtual) with RAM 8GB.
It does work in OSX 10.11.3.

In all the cases I've installed Node with *nvm*. Versions:
* node --version: *v5.9.1*
* npm --version: *3.7.3*

NOTE: In the failing host I've also tested with other versions of node+nmp.

**Comment by ibc** [comment]
I've applied the change in this [commit](https://github.com/npm/npm/pull/11947/files) and then, in the failing host I get this error:

> npm ERR! Linux 3.13-1-486
npm ERR! argv "/home/ibc/.nvm/versions/node/v5.9.1/bin/node" "/home/ibc/.nvm/versions/node/v5.9.1/bin/npm" "i"
npm ERR! node v5.9.1
npm ERR! npm  v3.7.3
>
npm ERR! Maximum call stack size exceeded

**Comment by ibc** [comment]
It works (in the same failing host as above) with node v4.4.1 and npm v2.14.20.

**Comment by mojodna** [comment]
@othiym23 following the same repro in https://github.com/npm/npm/issues/10686#issuecomment-183136413 with Node-5.9.1 and npm-3.8.5, the nested `npm install is-property` gives the following error:

$ npm install is-property
inherits@2.0.1 /private/tmp/npm/node_modules/inherits
└── is-property@1.0.2  extraneous

```
npm ERR! Darwin 15.4.0
npm ERR! argv "/Users/seth/.nvm/versions/node/v5.9.1/bin/node" "/Users/seth/.nvm/versions/node/v5.9.1/bin/npm" "install" "is-property"
npm ERR! node v5.9.1
npm ERR! npm  v3.8.5
npm ERR! path /private/tmp/npm/node_modules/inherits/node_modules/.staging/inherits-3fca3c0a
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/private/tmp/npm/node_modules/inherits/node_modules/.staging/inherits-3fca3c0a' -> '/private/tmp/npm/node_modules/inherits'
npm ERR! enoent ENOENT: no such file or directory, rename '/private/tmp/npm/node_modules/inherits/node_modules/.staging/inherits-3fca3c0a' -> '/private/tmp/npm/node_modules/inherits'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent
npm ERR! Darwin 15.4.0
npm ERR! argv "/Users/seth/.nvm/versions/node/v5.9.1/bin/node" "/Users/seth/.nvm/versions/node/v5.9.1/bin/npm" "install" "is-property"
npm ERR! node v5.9.1
npm ERR! npm  v3.8.5
npm ERR! path npm-debug.log.2829228988
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.2829228988'
npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.2829228988'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /private/tmp/npm/node_modules/inherits/npm-debug.log
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
```

**Comment by szarouski** [comment]
Steps to reproduce (Windows 10, node v4.1.0, npm v3.7.4)
npm i npm@3.7.4 -g
//npm -v === 3.7.4
npm i app-sandbox@2.x
cd node_modules/app-sandbox
npm i
[npm-debug.txt](https://github.com/npm/npm/files/190953/npm-debug.txt)

After upgrade to 3.8.3 I get another issue - `EBUSY: resource busy or locked`.
Steps to reproduce (in new folder):
npm i npm@3.8.3 -g
//npm -v === 3.8.3
npm i app-sandbox@2.x
cd node_modules/app-sandbox
npm i
[npm-debug-2.txt](https://github.com/npm/npm/files/190959/npm-debug-2.txt)

**Comment by nickbijmoer** [comment]
How to fix this? still get this issue : http://pastebin.com/sfs5YZVx

**Comment by syuilo** [comment]
I was solved by lowering the version.

**Comment by nickbijmoer** [comment]
Thx!

**Comment by roemhildtg** [comment]
Getting this error message today on a project:

https://gist.github.com/roemhildtg/5693a69eae662d34d8da7aa4f3b62ec6

I'm on OpenSuse x64 42.1 Leap

:~/public_html/cola/node_modules/bit-data-suite> npm --version3.7.3
~/public_html/cola/node_modules/bit-data-suite> node --version
v5.9.1

**Comment by jmwilkinson** [comment]
I want to caution people who use 3.8.3 and later.

Instead of getting that error, I got the following: 

`npm WARN [package] No repository field.
npm ERR! Linux 2.6.32-573.12.1.el6.x86_64
npm ERR! argv "/usr/bin/node" "/usr/local/bin/npm" "install" "request" "--save"
npm ERR! node v5.11.1
npm ERR! npm  v3.8.3
npm ERR! path [path]
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename [path] -> [path]
npm ERR! enoent ENOENT: no such file or directory, rename [path] -> [path]
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent
npm ERR! Linux 2.6.32-573.12.1.el6.x86_64
npm ERR! argv "/usr/bin/node" "/usr/local/bin/npm" "install" "request" "--save"
npm ERR! node v5.11.1
npm ERR! npm  v3.8.3
npm ERR! path npm-debug.log.3715500330
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.3715500330'
npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.3715500330'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     [log path]
`

The project I was trying to install a package for had been deleted. I replicated this multiple times with 3.8.3 and later. If you run into the `Cannot read property 'target' of null` error, please make sure you have an up-to-date backup of your project so you can restore should it be deleted. Obviously as the project was gone, I don't have a log file to review.

Version 3.3.12 of npm worked for me.

**Comment by fiws** [comment]
Eh Sorry for the reference spam above. I did not know that all these commits would land here.

Also: I am having some inconsistent behaviour if there is a `package.json` present and I use `--save` or not. Sometimes it loops, sometimes it hangs and sometimes it installs fine. racing condition?

Tested with my testpackage `fiws/test-npm-bug-10686`

**Comment by mattstyles** [comment]
I get the same repo deletion thing as @jmwilkinson 

```
npm ERR! Darwin 15.2.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "hogan.js"
npm ERR! node v6.2.1
npm ERR! npm  v3.9.3
npm ERR! path /usr/local/lib/node_modules/@dotlabel/temple/node_modules/.staging/@dotlabel/temple-9998f3cd
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/usr/local/lib/node_modules/@dotlabel/temple/node_modules/.staging/@dotlabel/temple-9998f3cd' -> '/usr/local/lib/node_modules/@dotlabel/temple'
npm ERR! enoent ENOENT: no such file or directory, rename '/usr/local/lib/node_modules/@dotlabel/temple/node_modules/.staging/@dotlabel/temple-9998f3cd' -> '/usr/local/lib/node_modules/@dotlabel/temple'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 
npm ERR! Darwin 15.2.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "hogan.js"
npm ERR! node v6.2.1
npm ERR! npm  v3.9.3
npm ERR! path npm-debug.log.2791326416
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.2791326416'
npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.2791326416'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/local/lib/node_modules/@dotlabel/temple/npm-debug.log
```

This was using `node -v 6.2.1` and  `npm -v 3.9.3`, slightly earlier versions threw the `target 'null'` errors reported earlier and `npm v3.3.12` worked.

**Comment by StarpTech** [comment]
Any updates?

**Comment by dominikwilkowski** [comment]
Same issue here:

```shell
$ npm i --save cli-size
npm ERR! Darwin 15.6.0
npm ERR! argv "/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/node" "/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/npm" "i" "--save" "cli-size"
npm ERR! node v6.7.0
npm ERR! npm  v3.10.3

npm ERR! Cannot read property 'cdl' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/dominikwilkowski/Sites/beast.js/npm-debug.log
```

Debug:
```shell
0 info it worked if it ends with ok
1 verbose cli [ '/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/node',
1 verbose cli   '/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/npm',
1 verbose cli   'i',
1 verbose cli   '--save',
1 verbose cli   'cli-size' ]
2 info using npm@3.10.3
3 info using node@v6.7.0
4 silly loadCurrentTree Starting
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 silly fetchPackageMetaData cli-size
8 silly fetchNamedPackageData cli-size
9 silly mapToRegistry name cli-size
10 silly mapToRegistry using default registry
11 silly mapToRegistry registry https://registry.npmjs.org/
12 silly mapToRegistry data Result {
12 silly mapToRegistry   raw: 'cli-size',
12 silly mapToRegistry   scope: null,
12 silly mapToRegistry   escapedName: 'cli-size',
12 silly mapToRegistry   name: 'cli-size',
12 silly mapToRegistry   rawSpec: '',
12 silly mapToRegistry   spec: 'latest',
12 silly mapToRegistry   type: 'tag' }
13 silly mapToRegistry uri https://registry.npmjs.org/cli-size
14 verbose request uri https://registry.npmjs.org/cli-size
15 verbose request no auth needed
16 info attempt registry request try #1 at 9:50:13 AM
17 verbose request id f1dcdfb12eabacf5
18 verbose etag "7XDSDCJ8UNIAG1YDJKX7OEKEF"
19 http request GET https://registry.npmjs.org/cli-size
20 http 304 https://registry.npmjs.org/cli-size
21 verbose headers { date: 'Tue, 04 Oct 2016 22:50:13 GMT',
21 verbose headers   via: '1.1 varnish',
21 verbose headers   'cache-control': 'max-age=300',
21 verbose headers   etag: '"7XDSDCJ8UNIAG1YDJKX7OEKEF"',
21 verbose headers   age: '0',
21 verbose headers   connection: 'keep-alive',
21 verbose headers   'x-served-by': 'cache-syd1621-SYD',
21 verbose headers   'x-cache': 'HIT',
21 verbose headers   'x-cache-hits': '1',
21 verbose headers   'x-timer': 'S1475621413.870967,VS0,VE900',
21 verbose headers   vary: 'Accept-Encoding' }
22 silly get cb [ 304,
22 silly get   { date: 'Tue, 04 Oct 2016 22:50:13 GMT',
22 silly get     via: '1.1 varnish',
22 silly get     'cache-control': 'max-age=300',
22 silly get     etag: '"7XDSDCJ8UNIAG1YDJKX7OEKEF"',
22 silly get     age: '0',
22 silly get     connection: 'keep-alive',
22 silly get     'x-served-by': 'cache-syd1621-SYD',
22 silly get     'x-cache': 'HIT',
22 silly get     'x-cache-hits': '1',
22 silly get     'x-timer': 'S1475621413.870967,VS0,VE900',
22 silly get     vary: 'Accept-Encoding' } ]
23 verbose etag https://registry.npmjs.org/cli-size from cache
24 verbose get saving cli-size to /Users/dominikwilkowski/.npm/registry.npmjs.org/cli-size/.cache.json
25 verbose correctMkdir /Users/dominikwilkowski/.npm correctMkdir not in flight; initializing
26 silly install normalizeTree
27 silly loadCurrentTree Finishing
28 silly loadIdealTree Starting
29 silly install loadIdealTree
30 silly cloneCurrentTree Starting
31 silly install cloneCurrentTreeToIdealTree
32 silly cloneCurrentTree Finishing
33 silly loadShrinkwrap Starting
34 silly install loadShrinkwrap
35 silly loadShrinkwrap Finishing
36 silly loadAllDepsIntoIdealTree Starting
37 silly install loadAllDepsIntoIdealTree
38 silly resolveWithNewModule cli-size@1.0.0 checking installable status
39 silly cache add args [ 'cli-size', null ]
40 verbose cache add spec cli-size
41 silly cache add parsed spec Result {
41 silly cache add   raw: 'cli-size',
41 silly cache add   scope: null,
41 silly cache add   escapedName: 'cli-size',
41 silly cache add   name: 'cli-size',
41 silly cache add   rawSpec: '',
41 silly cache add   spec: 'latest',
41 silly cache add   type: 'tag' }
42 silly addNamed cli-size@latest
43 verbose addNamed "latest" is being treated as a dist-tag for cli-size
44 info addNameTag [ 'cli-size', 'latest' ]
45 silly mapToRegistry name cli-size
46 silly mapToRegistry using default registry
47 silly mapToRegistry registry https://registry.npmjs.org/
48 silly mapToRegistry data Result {
48 silly mapToRegistry   raw: 'cli-size',
48 silly mapToRegistry   scope: null,
48 silly mapToRegistry   escapedName: 'cli-size',
48 silly mapToRegistry   name: 'cli-size',
48 silly mapToRegistry   rawSpec: '',
48 silly mapToRegistry   spec: 'latest',
48 silly mapToRegistry   type: 'tag' }
49 silly mapToRegistry uri https://registry.npmjs.org/cli-size
50 verbose addNameTag registry:https://registry.npmjs.org/cli-size not in flight; fetching
51 verbose get https://registry.npmjs.org/cli-size not expired, no request
52 silly addNameTag next cb for cli-size with tag latest
53 silly addNamed cli-size@1.0.0
54 verbose addNamed "1.0.0" is a plain semver version for cli-size
55 silly cache afterAdd cli-size@1.0.0
56 verbose afterAdd /Users/dominikwilkowski/.npm/cli-size/1.0.0/package/package.json not in flight; writing
57 verbose correctMkdir /Users/dominikwilkowski/.npm correctMkdir not in flight; initializing
58 verbose afterAdd /Users/dominikwilkowski/.npm/cli-size/1.0.0/package/package.json written
59 verbose stack TypeError: Cannot read property 'cdl' of null
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/install/deps.js:635:21
59 verbose stack     at Array.some (native)
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/install/deps.js:634:43
59 verbose stack     at Array.some (native)
59 verbose stack     at exports.earliestInstallable (/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/install/deps.js:631:68)
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/install/deps.js:533:20
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/node_modules/iferr/index.js:13:50
59 verbose stack     at addBundled (/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/fetch-package-metadata.js:245:39)
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/install/deps.js:532:5
59 verbose stack     at /Users/dominikwilkowski/.nvm/versions/node/v6.7.0/lib/node_modules/npm/node_modules/iferr/index.js:13:50
60 verbose cwd /Users/dominikwilkowski/Sites/beast.js
61 error Darwin 15.6.0
62 error argv "/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/node" "/Users/dominikwilkowski/.nvm/versions/node/v6.7.0/bin/npm" "i" "--save" "cli-size"
63 error node v6.7.0
64 error npm  v3.10.3
65 error Cannot read property 'cdl' of null
66 error If you need help, you may report this error at:
66 error     <https://github.com/npm/npm/issues>
67 verbose exit [ 1, true ]
```

**Comment by IonicaBizau** [comment]
This is fixed in #13999. 🚀 

In https://github.com/Bloggify/starter/issues/2 I put together the instructions how to install my patched version.

**Comment by ackalker** [comment]
@IonicaBizau and others:
Before everyone goes off to the celebrations and pops open the champagne, could someone please explain why the bad commit in [my bisection attempt](https://github.com/npm/npm/issues/10686#issuecomment-200353028) causes the original problem? I suspect there's more to this bug than just the proposed fix, namely an incomplete or faulty rename / refactoring.

**Comment by IonicaBizau** [comment]
@ackalker Actually, it turns out this is a similar issue, but I don't think it's the same with https://github.com/npm/npm/issues/13997 (the one I fixed).

**Comment by ackalker** [comment]
@IonicaBizau Ah, okay, thanks for the quick reaction!

**Comment by michaelfakhri** [comment]
I recently faced similar issues when trying install @browser-ipfs. I also regularly screw up the ownership of my system files, so I tried:
 sudo npm install .... 
It seems to have avoided this error or fixed whatever the cause is (be it ownership related or not).
Hope this helps anyone who's facing the same issue.

**Comment by She110ck** [comment]
It's worked for me:
1. `sudo npm i npm@latest -g`
2. continue work on  **new** command line
  
Ubuntu 16.04x64

**Comment by hanishassim** [comment]
Your solution works for me @She110ck, thanks!

**Comment by shimu12** [comment]
Your solution works for me @She110ck, thanks!

**Comment by puneetdewan** [comment]
Guys I am trying to install this package. But it is giving an error> Help me out:?

C:\Users\puneet.dewan>node -v
v7.4.0

C:\Users\puneet.dewan>npm -v
4.0.5

C:\Users\puneet.dewan>npm install -g babel
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "babel"
npm ERR! node v7.4.0
npm ERR! npm  v4.0.5

npm ERR! Cannot read property 'path' of null
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "babel"
npm ERR! node v7.4.0
npm ERR! npm  v4.0.5
npm ERR! code ECONNRESET

npm ERR! network socket hang up
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\puneet.dewan\npm-debug.log

**Comment by puneetdewan** [comment]
# npm-debug.log

0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   '-g',
1 verbose cli   'babel' ]
2 info using npm@4.0.5
3 info using node@v7.4.0
4 silly loadCurrentTree Starting
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 silly fetchPackageMetaData babel
8 silly fetchNamedPackageData babel
9 silly mapToRegistry name babel
10 silly mapToRegistry using default registry
11 silly mapToRegistry registry https://registry.npmjs.org/
12 silly mapToRegistry data Result {
12 silly mapToRegistry   raw: 'babel',
12 silly mapToRegistry   scope: null,
12 silly mapToRegistry   escapedName: 'babel',
12 silly mapToRegistry   name: 'babel',
12 silly mapToRegistry   rawSpec: '',
12 silly mapToRegistry   spec: 'latest',
12 silly mapToRegistry   type: 'tag' }
13 silly mapToRegistry uri https://registry.npmjs.org/babel
14 verbose request uri https://registry.npmjs.org/babel
15 verbose request no auth needed
16 info attempt registry request try #1 at 3:22:55 PM
17 verbose request id 95d3d5ccc39406c8
18 http request GET https://registry.npmjs.org/babel
19 info retry will retry, error on last attempt: Error: tunneling socket could not be established, cause=connect ETIMEDOUT 74.122.238.10:80
20 info attempt registry request try #2 at 3:23:26 PM
21 http request GET https://registry.npmjs.org/babel
22 info retry will retry, error on last attempt: Error: tunneling socket could not be established, cause=connect ETIMEDOUT 74.122.238.10:80
23 info attempt registry request try #3 at 3:24:47 PM
24 http request GET https://registry.npmjs.org/babel
25 silly fetchPackageMetaData Error: tunneling socket could not be established, cause=connect ETIMEDOUT 74.122.238.10:80
25 silly fetchPackageMetaData     at ClientRequest.onError (C:\Program Files\nodejs\node_modules\npm\node_modules\request\node_modules\tunnel-agent\index.js:176:17)
25 silly fetchPackageMetaData     at Object.onceWrapper (events.js:290:19)
25 silly fetchPackageMetaData     at emitOne (events.js:96:13)
25 silly fetchPackageMetaData     at ClientRequest.emit (events.js:188:7)
25 silly fetchPackageMetaData     at Socket.socketErrorListener (_http_client.js:309:9)
25 silly fetchPackageMetaData     at emitOne (events.js:96:13)
25 silly fetchPackageMetaData     at Socket.emit (events.js:188:7)
25 silly fetchPackageMetaData     at emitErrorNT (net.js:1281:8)
25 silly fetchPackageMetaData     at _combinedTickCallback (internal/process/next_tick.js:74:11)
25 silly fetchPackageMetaData     at process._tickCallback (internal/process/next_tick.js:98:9)
25 silly fetchPackageMetaData  error for babel { Error: tunneling socket could not be established, cause=connect ETIMEDOUT 74.122.238.10:80
25 silly fetchPackageMetaData     at ClientRequest.onError (C:\Program Files\nodejs\node_modules\npm\node_modules\request\node_modules\tunnel-agent\index.js:176:17)
25 silly fetchPackageMetaData     at Object.onceWrapper (events.js:290:19)
25 silly fetchPackageMetaData     at emitOne (events.js:96:13)
25 silly fetchPackageMetaData     at ClientRequest.emit (events.js:188:7)
25 silly fetchPackageMetaData     at Socket.socketErrorListener (_http_client.js:309:9)
25 silly fetchPackageMetaData     at emitOne (events.js:96:13)
25 silly fetchPackageMetaData     at Socket.emit (events.js:188:7)
25 silly fetchPackageMetaData     at emitErrorNT (net.js:1281:8)
25 silly fetchPackageMetaData     at _combinedTickCallback (internal/process/next_tick.js:74:11)
25 silly fetchPackageMetaData     at process._tickCallback (internal/process/next_tick.js:98:9) code: 'ECONNRESET' }
26 silly rollbackFailedOptional Starting
27 verbose stack TypeError: Cannot read property 'path' of null
27 verbose stack     at Installer.rollbackFailedOptional (C:\Program Files\nodejs\node_modules\npm\lib\install.js:505:29)
27 verbose stack     at Array.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
27 verbose stack     at LOOP (C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
27 verbose stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\chain.js:18:7
27 verbose stack     at C:\Program Files\nodejs\node_modules\npm\lib\install.js:331:5
27 verbose stack     at Array.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
27 verbose stack     at LOOP (C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
27 verbose stack     at chain (C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\chain.js:20:5)
27 verbose stack     at C:\Program Files\nodejs\node_modules\npm\lib\install.js:288:5
27 verbose stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\slide\lib\chain.js:16:22
28 verbose cwd C:\Users\puneet.dewan
29 error Windows_NT 10.0.14393
30 error argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "babel"
31 error node v7.4.0
32 error npm  v4.0.5
33 error Cannot read property 'path' of null
34 error If you need help, you may report this error at:
34 error     <https://github.com/npm/npm/issues>
35 verbose exit [ 1, true ]

**Comment by iarna** [comment]
This is a very old issue and likely fixed now. Further, the most recent error logs in it are not related to the original issue, so I'm gonna close this. If you see anything like this with npm@5 please open a new issue!
